docs/auth-code-pkce-grant.md

```diff
@@ -0,0 +1,168 @@
+# Auth Code Grant 
+
+## Usages
+
+The OAuth 2.0 Proof Key for Code Exchange flow is described in [RFC 7636](https://tools.ietf.org/html/rfc7636). It is used to perform authentication and authorization typically in server-side web applications where the client_secret can be securely stored and kept safe. The flow enables apps to securely acquire access_tokens that can be used to access secured resources. 
+More BCP (as of 20190513) discussion in this draft: https://tools.ietf.org/html/draft-ietf-oauth-browser-based-apps-00
+
+## Flow Diagram
+
+<img src="./AuthCodePKCEGrantDiagram.png" />
+
+<!-- Paste the following comment in this web site to edit the diagram and get a new image https://www.websequencediagrams.com/ -->
+<!--
+Flow Sequence:
+Native App->/auth: Pops a browser dialog.\nSends random State and sha256 Challenge of Verifier. \nRequests AuthCode.
+Native App->/auth: User completes policy
+/auth->Native App: Returns AuthCode and State
+Native App->/token: Verifies State, sends AuthCode, Verifier and requests OAuth tokens
+/token->Native App: Returns an access token and a refresh token
+Native App->/api:Calls API with Access token in Authorization header
+/api->/api:Validate Access token
+/api->Native App:Returns data to app
+alt Access token expires
+end
+Native App->/token: Request a new token using refresh_token
+/token->Native App: Returns a new access token
+Native App->/api:Calls API with access token in Authorization header
+-->
+
+
+## Request Authorization Code
```

Comment:
RFC states that along with the `code_challenge` there is an optional [`code_challenge_method`](https://tools.ietf.org/html/rfc7636#section-4.3)

>       OPTIONAL, **defaults to "plain" if not present in the request.**  Code verifier transformation method is "S256" or "plain".

Our system is defaulting to S256 and I'm ok with it not supporting 'plain', but we shouldn't change the RFC declared default. We should document this so that a `code_challenge_method` value of 'S256' is required.

On the server I would expect us to check that and throw an exception if the value isn't 'S256'.

Reply:
Great idea. 

I updated the html and md here and also fixed the server code.